dapr/proto/components/v1/pubsub.proto

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+package dapr.proto.components.v1;
+
+import "dapr/proto/components/v1/common.proto";
+
+option go_package = "github.com/dapr/dapr/pkg/proto/components/v1;components";
+
+// PubSub service provides a gRPC interface for pubsub components.
+service PubSub {
+  // Initializes the pubsub component with the given metadata.
+  rpc Init(PubSubInitRequest) returns (PubSubInitResponse) {}
+
+  // Returns a list of implemented pubsub features.
+  rpc Features(FeaturesRequest) returns (FeaturesResponse) {}
+
+  // Publish publishes a new message for the given topic.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+
+  // Establishes a stream with the server, which sends messages down to the
+  // client. The client streams acknowledgements back to the server. The server
+  // will close the stream and return the status on any error. In case of closed
+  // connection, the client should re-establish the stream.
+  // The message MUST contain a `subscription` on it that should be used for the
+  // entire streaming pull.
+  rpc PullMessages(stream PullMessagesRequest)
+      returns (stream PullMessageResponse) {}
+
+  // Ping the pubsub. Used for liveness porpuses.
+  rpc Ping(PingRequest) returns (PingResponse) {}
+}
+
+// Used for describing errors when ack'ing messages.
+message AckMessageError {
+  string message = 1;
+}
+
+// Used for acknowledge a message.
+message PullMessagesRequest {
+  // Required. The subscription for which to initialize the new stream. This
```

Comment:
They are indeed two different problems, using `oneof` or not will not solve the required/optional problem, the comment will continue to be necessary, in fact all fields are optional (since the version 3 the idea of having required and optional fields ended up being deprecated https://github.com/protocolbuffers/protobuf/issues/2497).

Regard the `oneof` it was my first design but I decided to remove - however I can revisit this decision again, no worries, but let's go to the facts

1) oneof will have almost the same semantics as comment + optional attribute as the subscription will be ignored for the subsequent messages, so it will only make the sdks implementations more difficult and I believe it will make it more confusing, even though it seems clearer
2) attributes are easier to handle and allowing defining both at same time is not so bad - we can leverage this in the future to define attributes that can be modified on-the-fly in the subscription


To summarize:

I think oneof is good in the sense of removing the possibility of specified both, ack and subscription, but the real problem here is order, Subscription should come first than acks (obviously) and the oneof will end up making implementation difficult due to type checking

But again - I can revisit this decision